evalml/tests/model_understanding_tests/prediction_explanations_tests/test_explainers.py

```diff
@@ -1148,7 +1148,20 @@ def test_json_serialization(
         num_to_explain=1,
         output_format="dict",
     )
-    assert json.loads(json.dumps(best_worst)) == best_worst
+
+    def str_to_num_decoder(dct):
```

Comment:
I'm concerned by this change as well. I think `json.loads(json.dumps(best_worst)) == best_worst` is an axiom we should maintain, rather than patching it over for the sake of the upgrade. What change predicated this, and how can we account for it in prediction explanations instead?

Reply:
This is because the output of `explain_predictions_best_worst` results in the keys of `predicted_values` being integers, since they're numeric values being represented as strings. Woodwork's new inference will infer strings as their appropriate numeric logical type if their incoming dtype is `object` and no other logical type is specified.
```
'predicted_values': {'error_name': 'Cross Entropy',
                                   'error_value': 9.967001210947085e-07,
                                   'index_id': 33,
                                   'predicted_value': 1,
                                   'probabilities': {0: 0.0, 1: 1.0},
                                   'target_value': 1},
```
but JSON doesn't allow non-string keys, so it replaces them with
```
'predicted_values': {'error_name': 'Cross Entropy',
                                   'error_value': 9.967001210947085e-07,
                                   'index_id': 33,
                                   'predicted_value': 1,
                                   'probabilities': {'0': 0.0, '1': 1.0}, # String keys
                                   'target_value': 1},
```
What I added here was an `object_hook` to be called instead of the standard `dict` which converts the `string` key back to an `int` just so they can be compared for the test.

I think what's happening is that [this](https://github.com/alteryx/evalml/blob/main/evalml/model_understanding/prediction_explanations/_user_interface.py#L757) is creating keys and assigning them to the `probabilities` key even if they're integers. I can modify this to stringify all keys since JSON will end up converting them anyways. I just thought it would be better for the output of `explain_predictions_best_worst` to represent the actual values even if JSON would end up stringifying them. @eccabay @jeremyliweishih does that make sense?

Reply:
wait @ParthivNaresh why does woodwork inference matter when we're testing if dumping and loading the same json will result in the same dictionary?

Reply:
@jeremyliweishih Because JSON only allows key names to be strings, so if the values themselves aren't strings, then they get dumped as integers and read as strings. I think somewhere in the prediction explanation process, the original dtype of `y` is lost and reinferred, which means if the original dtype was a string, it would be reinferred as an integer.

I tried replacing 
```
elif problem_type == problem_type.BINARY:
        X, y = X_y_binary
        y = pd.Series(y).astype("str")
        pipeline = logistic_regression_binary_pipeline
```
with
```
elif problem_type == problem_type.BINARY:
        X, y = X_y_binary
        y = ww.init_series(pd.Series(y), logical_type="Unknown")
        pipeline = logistic_regression_binary_pipeline
```
or explicitly setting it as string dtype (`string` instead of `str`), but it still writes out the dictionary in `best_worst` as an integer

Reply:
> I just thought it would be better for the output of explain_predictions_best_worst to represent the actual values even if JSON would end up stringifying them
Doesn't your second comment contradict this, so `explain_predictions_best_worst` is still modifying the actual values, but in the wrong direction? 

My preference remains that we keep the assertion that `json.loads(json.dumps(best_worst)) == best_worst`, and we update the code to maintain this rather than the test.

Reply:
`explain_predictions_best_worst` is modifying this somewhere but that's not the expectation. If I pass in a string dtype (which Woodwork won't infer as numeric, but Categorical in this case) then that would be the expected output.

I'm taking a look into `explain_predictions_best_worst` to see where this unexpected reinference is happening

Reply:
@eccabay I think the issue was that `ww.init_series` was being used to determine the unique classes that existed in a classification pipeline. If the column names were strings then they got inferred as integers because of this.